### PR TITLE
planner, executor: support insert/replace into partition

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?
Before this commit:
``` sql
tidb> insert into t1 partition (p0,p1) values(2, 'a');
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 24 near "partition (p0,p1) values(2, 'a')"
```

After this commit:
``` sql
tidb> insert into t1 partition (p0,p1) values(2, 'a');
Query OK, 1 row affected (0.00 sec)

tidb> select * from t1 partition(p0);
+------+------+
| a    | b    |
+------+------+
|    2 | a    |
+------+------+
1 row in set (0.00 sec)
```

Problem Summary:

### What is changed and how it works?
Check the given partition set when executing insert/replace.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

- Performance regression
    - Consumes more CPU

### Release note 
- Support `insert/replace into tbl_name partition(partition_name_list)`

### CODE REVIEW

File: executor/insert_common.go [hunk 1]
```diff
@@ -217,13 +218,24 @@ func insertRows(ctx context.Context, base insertCommon) (err error) {
 	rows := make([][]types.Datum, 0, len(e.Lists))
 	memUsageOfRows := int64(0)
 	memTracker := e.memTracker
+	var partitionTable table.PartitionedTable
+	if len(e.Partitions) != 0 {
+		partitionTable = e.Table.(table.PartitionedTable)
+		if len(e.Partitions) != 0 {
```
Comment (imtbkcat) [inline executor/insert_common.go:224]: What does this `if` brach means?

File: executor/executor_test.go [hunk 2]
```diff
@@ -5561,3 +5561,88 @@ func (s *testSuite1) TestInsertValuesWithSubQuery(c *C) {
...
+	err = tk.ExecToErr("replace into t1 partition(p0, p_non_exist) values(1, 'a')")
+	c.Assert(err.Error(), Equals, "[table:1735]Unknown partition 'p_non_exist' in table 't1'")
+
+	err = tk.ExecToErr("replace into t1 partition(p0, p1) values(40, 'a')")
+	c.Assert(err.Error(), Equals, "[executor:1748]Found a row not matching the given partition set")
+
+	tk.MustExec("truncate table t1")
+
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a int, b char(10))")
```
Comment (imtbkcat) [inline executor/executor_test.go:5611]: add a test to test like `insert into t partition(p0, p1) values(1, 'a')` here `t` is a general table.

File: executor/insert_common.go [hunk 3]
```diff
@@ -262,6 +271,20 @@ func insertRows(ctx context.Context, base insertCommon) (err error) {
 	return nil
 }
 
+func checkRowDoesNotMatchGivenPartitionSet(e *InsertValues, partitionTable table.PartitionedTable, row []types.Datum) error {
+	if partitionTable == nil {
+		return nil
+	}
+	p, err := partitionTable.GetPartitionByRow(e.ctx, row)
```
Comment (tiancaiamao) [inline executor/insert_common.go:278]: This check is a kind of too heavy, you may see the implementation of `GetPartitionByRow`.
We already have some place calling `GetPartitionByRow`, could it be moved there?

**Comment by XuHuaiyu** [comment]
wait for https://github.com/pingcap/tidb/pull/17285

**Comment by imtbkcat** [comment]
@XuHuaiyu please update this pr.

File: executor/insert_common.go [hunk 4]
```diff
@@ -71,6 +71,7 @@ type InsertValues struct {
 	// https://dev.mysql.com/doc/refman/8.0/en/innodb-auto-increment-handling.html
 	lazyFillAutoID bool
 	memTracker     *memory.Tracker
+	Partitions     map[int64]struct{}
```
Comment (tiancaiamao) [inline executor/insert_common.go:74]: not used?

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**